crates/bevy_core_pipeline/src/mipmap_generator.rs

```diff
@@ -0,0 +1,602 @@
+//! A wrapper that runs a shader repeatedly to populate a series of mipmaps.
+//!
+//! A variety of passes need to or will need to do this, and using this module
+//! helps to reduce boilerplate and code duplication.
+
+use std::{borrow::Cow, hash::Hash, marker::PhantomData};
+
+use bevy_app::{App, Plugin};
+use bevy_asset::Handle;
+use bevy_ecs::{
+    component::Component,
+    system::Resource,
+    world::{FromWorld, World},
+};
+use bevy_math::UVec2;
+use bevy_render::{
+    render_resource::{
+        AddressMode, BindGroup, BindGroupEntry, BindGroupLayout, BindGroupLayoutDescriptor,
+        BindGroupLayoutEntry, BindingResource, BindingType, CachedRenderPipelineId,
+        ColorTargetState, ColorWrites, Extent3d, FilterMode, FragmentState, LoadOp,
+        MultisampleState, Operations, PipelineCache, PrimitiveState, RenderPassColorAttachment,
+        RenderPassDescriptor, RenderPipelineDescriptor, Sampler, SamplerBindingType,
+        SamplerDescriptor, Shader, ShaderDefVal, ShaderStages, SpecializedRenderPipeline,
+        SpecializedRenderPipelines, TextureDescriptor, TextureDimension, TextureFormat,
+        TextureSampleType, TextureUsages, TextureView, TextureViewDescriptor, TextureViewDimension,
+    },
+    renderer::{RenderContext, RenderDevice},
+    texture::{CachedTexture, TextureCache},
+    RenderApp,
+};
+
+use crate::fullscreen_vertex_shader::fullscreen_shader_vertex_state;
+
+/// All mipmappers implement this trait.
+///
+/// The type that implements this trait is used as the pipeline key for the mipmapping shader.
+pub trait Mipmap: Clone + PartialEq + Eq + Hash + Send + Sync + 'static {
```

Comment:
I have a few comments about this:

- I think both `add_custom_*` functions should return arrays instead of passing a mutable reference. Especially since it's named `add_`.
- I also think the `add_custom_*` part should be removed. Just `shader_defs` should be enough, but that's a pretty subjective.
- I haven't fully read everything so I might be missing something, but would it be possible to use AsBindGroup instead of the `add_custom_bind_group_layout_enties`. This api is very similar to the Material api so would be nice if it used something similar for that too.